// @pm2/io/build/main/src/services/inspector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const inspector = require("inspector");
const debug_1 = require("debug");
const debug = debug_1.default('axm:inspectorservice');
class InspectorService {
    constructor() {
        this.isConnected = false;
    }
    createSession() {
        if (!this.session) {
            this.session = new inspector.Session();
        }
        return this.session;
    }
    post(action, params) {
        return new Promise((resolve, reject) => {
            this.session.post(action, params, (err, data) => {
                if (err)
                    return reject(err);
                debug(action + ' !');
                resolve(data);
            });
        });
    }
    on(eventName, callback) {
        this.session.on(eventName, callback);
    }
    connect() {
        if (!this.isConnected) {
            this.session.connect();
        }
        this.isConnected = true;
    }
    disconnect() {
        if (this.isConnected) {
            this.session.post('Profiler.disable');
            this.session.disconnect();
            this.isConnected = false;
        }
        else {
            debug('No open session !');
        }
    }
}
exports.InspectorService = InspectorService;
module.exports = new InspectorService();
