// @pm2/io/build/main/src/services/transport.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const debug_1 = require("debug");
const semver = require("semver");
const stringify = require("json-stringify-safe");
const debug = debug_1.default('axm:transportService');
class TransportConfig {
}
exports.TransportConfig = TransportConfig;
class Actions {
}
exports.Actions = Actions;
class Process {
}
exports.Process = Process;
class Transport {
}
exports.Transport = Transport;
class Agent {
}
exports.Agent = Agent;
class TransportService {
    constructor() {
        this.isStandalone = false;
        this.initiated = false; // tslint:disable-line
    }
    init() {
        this.initiated = true;
        this.isStandalone = false;
    }
    initStandalone(config) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!semver.satisfies(process.version, '>= 6.0.0')) {
                this.init();
                console.error('[STANDALONE MODE] Unable to set standalone mode with node < 6.0.0');
                return process.exit(1);
            }
            const AgentNode = require('@pm2/agent-node');
            this.isStandalone = true;
            this.initiated = true;
            debug('Init new transport service');
            this.config = config;
            this.process = {
                axm_actions: [],
                axm_options: {},
                axm_monitor: {}
            };
            this.agent = new AgentNode({
                publicKey: this.config.publicKey,
                secretKey: this.config.secretKey,
                appName: this.config.appName,
                serverName: this.config.serverName
            }, this.process);
            this.agent.sendLogs = config.sendLogs || false;
            try {
                yield this.agent.start();
            }
            catch (err) {
                console.error(err);
                process.exit(1);
            }
            this.transport = this.agent.transport;
            return debug('Agent launched');
        });
    }
    setMetrics(metrics) {
        if (this.isStandalone) {
            return this.process.axm_monitor = metrics;
        }
        this.send('axm:monitor', metrics);
    }
    addAction(action) {
        debug(`Add action: ${action.action_name}:${action.action_type}`);
        if (this.isStandalone) {
            return this.process.axm_actions.push(action);
        }
        return this.send('axm:action', action);
    }
    setOptions(options) {
        debug(`Set options: [${Object.keys(options).join(',')}]`);
        if (this.isStandalone) {
            return this.process.axm_options = Object.assign(this.process.axm_options, options);
        }
        return this.send('axm:option:configuration', options);
    }
    getFormattedPayload(channel, payload) {
        // Reformat for backend
        switch (channel) {
            case 'axm:reply':
                return { data: payload };
            case 'process:exception':
                return { data: payload };
        }
        return payload;
    }
    send(channel, payload) {
        if (this.isStandalone) {
            return this.agent.send(channel, this.getFormattedPayload(channel, payload)) ? 0 : -1;
        }
        if (!process.send)
            return -1;
        try {
            process.send(JSON.parse(stringify({
                type: channel,
                data: payload
            })));
        }
        catch (e) {
            debug('Process disconnected from parent !');
            debug(e.stack || e);
            return process.exit(1);
        }
        return 0;
    }
    destroy() {
        if (!this.isStandalone)
            return;
        this.transport.disconnect();
    }
}
exports.default = TransportService;
