// @pm2/io/build/main/src/features/profiling.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const profilingCPUFallback_1 = require("../profiling/profilingCPUFallback");
const profilingHeapFallback_1 = require("../profiling/profilingHeapFallback");
const configuration_1 = require("../configuration");
const semver = require("semver");
class ProfilingFeature {
    init(forceFallback) {
        // allow to force the fallback via environment
        if (process.env.PM2_PROFILING_FORCE_FALLBACK)
            forceFallback = true;
        const isInspectorOk = (semver.satisfies(process.version, '>= 10.0.0') ||
            (semver.satisfies(process.version, '>= 8.0.0') && process.env.FORCE_INSPECTOR)) && !forceFallback;
        let ProfilingCPU;
        let ProfilingHeap;
        if (isInspectorOk) {
            ProfilingCPU = require('../profiling/profilingCPU').default;
            ProfilingHeap = require('../profiling/profilingHeap').default;
            configuration_1.default.configureModule({
                heapdump: true
            });
        }
        this.profilings = {
            cpuProfiling: isInspectorOk ? new ProfilingCPU() : new profilingCPUFallback_1.default(),
            heapProfiling: isInspectorOk ? new ProfilingHeap() : new profilingHeapFallback_1.default()
        };
        return this.profilings;
    }
    destroy() {
        for (let profilingName in this.profilings) {
            if (typeof this.profilings[profilingName].destroy === 'function') {
                this.profilings[profilingName].destroy();
            }
        }
    }
}
exports.default = ProfilingFeature;
