// @pm2/io/build/main/src/metrics/deepMetrics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const metricConfig_1 = require("../utils/metricConfig");
const deepMetricsTracer_1 = require("./deepMetricsTracer");
const deepmetrics = require("deep-metrics");
const debug_1 = require("debug");
const debug = debug_1.default('axm:deepMetrics');
class DeepMetrics {
    constructor(metricFeature) {
        this.allPossibleMetrics = {};
        this.defaultConf = {
            mongo: true,
            mysql: true,
            mqtt: true,
            socketio: true,
            redis: true,
            http: true,
            https: true,
            'http-outbound': true,
            'https-outbound': true
        };
        this.metricFeature = metricFeature;
    }
    init(config) {
        deepmetrics.start();
        // instantiate all metrics
        for (let probeName in this.defaultConf) {
            if (this.defaultConf.hasOwnProperty(probeName)) {
                this.allPossibleMetrics[probeName] = new deepMetricsTracer_1.default(this.metricFeature, deepmetrics.ee, probeName);
            }
        }
        config = metricConfig_1.default.getConfig(config, this.defaultConf);
        // initialize only metrics found in config
        for (let probeName in this.allPossibleMetrics) {
            if (this.allPossibleMetrics.hasOwnProperty(probeName) && (config === 'all' || config[probeName] === true)) {
                this.allPossibleMetrics[probeName].init();
            }
        }
    }
    destroy() {
        deepmetrics.stop();
        // clean children
        for (let probeName in this.allPossibleMetrics) {
            if (this.allPossibleMetrics.hasOwnProperty(probeName)) {
                this.allPossibleMetrics[probeName].destroy();
            }
        }
        this.allPossibleMetrics = {};
        debug('Deep metrics detroyed !');
    }
}
exports.default = DeepMetrics;
