// @pm2/io/build/main/src/metrics/eventLoopHandlesRequests.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const debug_1 = require("debug");
const debug = debug_1.default('axm:eventLoop');
class EventLoopHandlesRequestsMetric {
    constructor(metricFeature) {
        this.metricFeature = metricFeature;
    }
    init(config) {
        if (typeof this.getProcess()._getActiveRequests === 'function') {
            this.metricFeature.metric({
                name: 'Active requests',
                type: 'internal/libuv/requests',
                value: () => { return this.getProcess()._getActiveRequests().length; }
            });
        }
        if (typeof this.getProcess()._getActiveHandles === 'function') {
            this.metricFeature.metric({
                name: 'Active handles',
                type: 'internal/libuv/handles',
                value: () => { return this.getProcess()._getActiveHandles().length; }
            });
        }
    }
    destroy() {
        debug('EventLoopHandlesRequestsMetric destroyed !');
    }
    getProcess() {
        return process;
    }
}
exports.default = EventLoopHandlesRequestsMetric;
