// @pm2/io/build/main/src/metrics/metricFromDump.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const miscellaneous_1 = require("../utils/miscellaneous");
const debug_1 = require("debug");
const debug = debug_1.default('axm:metricfromdump');
class MetricFromDump {
    constructor(metricFeature, conf) {
        this.conf = Array.isArray(conf) ? conf : [conf];
        this.metricFeature = metricFeature;
    }
    init() {
        const self = this;
        for (let i = 0; i < this.conf.length; i++) {
            (function (index) {
                self.metricFeature.metric({
                    name: self.conf[index].name,
                    type: `internal/libuv/${self.conf[index].name}`,
                    value: () => {
                        return miscellaneous_1.default.getValueFromDump(self.conf[index].property, self.conf[index].parentProperty);
                    }
                });
            })(i);
        }
    }
    destroy() {
        debug(`${this.conf.name} metric destroyed !`);
    }
}
exports.default = MetricFromDump;
